safehttp/form.go

```diff
@@ -0,0 +1,222 @@
+// values. In case it does, it will try to convert the first valid
+// uint64 and return it. If there are no values associated with
+// param, it will return the default value. If the first value is not an
+// unsigned integer, it will return the default value and set the Form
+// error field.
+func (f *Form) Uint64(param string, defaultValue uint64) uint64 {
+	vals, ok := f.values[param]
+	if !ok {
+		return defaultValue
+	}
+	paramVal, err := strconv.ParseUint(vals[0], 10, 64)
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+
+// String checks whether key param maps to any query or form parameter
+// values. In case it does, it will return the first value. If it doesn't, it
+// will return the default value.
+func (f *Form) String(param string, defaultValue string) string {
+	vals, ok := f.values[param]
+	if !ok {
+		return defaultValue
+	}
+	return vals[0]
+}
+
+// Float64 checks whether key param maps to any query or form parameter
+// values. In case it does, it will try to convert the first value to a valid
+// float64 and return it. If there are no values associated with param, it will
+// return the default value. If the first value is not a float, it will return
+// the default value and Err() will return the parsing error.
+func (f *Form) Float64(param string, defaultValue float64) float64 {
+	vals, ok := f.values[param]
+	if !ok {
+		return defaultValue
+	}
+	paramVal, err := strconv.ParseFloat(vals[0], 64)
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+
+// Bool checks whether key param maps to any query or form parameter
+// values. In case it does, it will try to convert the first value to a valid
+// bool and return it. If there are no values associated with param, it will
+// return the default value. If the first value is not a boolean, it will return
+// the default value and Err() will return the parsing error.
+func (f *Form) Bool(param string, defaultValue bool) bool {
+	vals, ok := f.values[param]
+	if !ok {
+		return defaultValue
+	}
+	switch vals[0] {
+	case "true":
+		return true
+	case "false":
+		return false
+	default:
+		f.err = fmt.Errorf("values of form parameter %q not a boolean", param)
+	}
+	return false
+}
+
+func clearSlice(slicePtr interface{}) error {
+	switch vs := slicePtr.(type) {
+	case *[]string:
+		*vs = nil
+	case *[]int64:
+		*vs = nil
+	case *[]float64:
+		*vs = nil
+	case *[]uint64:
+		*vs = nil
+	case *[]bool:
+		*vs = nil
+	default:
+		return fmt.Errorf("type not supported in Slice call: %T", vs)
+	}
+	return nil
+}
+
+// Slice checks whether key param maps to any query or form parameters. If it
+// does, it will try to convert them to the type of slice elements slicePtr
+// points to. If there are no values associated with param, it will clear the
+// slice. If type conversion fails at any point, the Form error field will be
+// set and the slice will be cleared.
+func (f *Form) Slice(slicePtr interface{}, param string) {
```

Comment:
IMHO it looks better to do `f.Slice("names", &s)` rather than `f.Slice(&s, "names")`, so you might consider reversing the order.

Reply:
You're right, thanks :)

Reply:
Missed.